subcommand/connect-init/command_test.go

```diff
@@ -40,18 +61,249 @@ func TestRun_FlagValidation(t *testing.T) {
+		"-skip-service-registration-polling=false"}
+	// Run the command.
+	code := cmd.Run(flags)
+	require.Equal(t, 0, code)
+
+	// Validate contents of proxyFile.
+	data, err := ioutil.ReadFile(proxyFile)
+	require.NoError(t, err)
+	require.Contains(t, string(data), "counting-counting-sidecar-proxy")
+}
+
+// This test validates happy path without ACLs : wait on proxy+service to be registered and write out proxyid file
+func TestRun_happyPathNoACLs(t *testing.T) {
+	t.Parallel()
+	// This is the output file for the proxyid.
+	proxyFile := common.WriteTempFile(t, "")
+
+	// Start Consul server.
+	server, err := testutil.NewTestServerConfigT(t, nil)
+	defer server.Stop()
+	require.NoError(t, err)
+	server.WaitForLeader(t)
+	consulClient, err := api.NewClient(&api.Config{Address: server.HTTPAddr})
+	require.NoError(t, err)
+
+	// Register Consul services.
+	testConsulServices := []api.AgentServiceRegistration{consulCountingSvc, consulCountingSvcSidecar}
+	for _, svc := range testConsulServices {
+		require.NoError(t, consulClient.Agent().ServiceRegister(&svc))
+	}
+
+	ui := cli.NewMockUi()
+	cmd := Command{
+		UI:           ui,
+		consulClient: consulClient,
+		ProxyIDFile:  proxyFile,
+	}
+	code := cmd.Run(defaultTestFlags)
+	require.Equal(t, 0, code)
+
+	// Validate contents of proxyFile.
+	data, err := ioutil.ReadFile(proxyFile)
+	require.NoError(t, err)
+	require.Contains(t, string(data), "counting-counting-sidecar-proxy")
+}
+
+// TestRun_RetryServicePolling starts the command and does not register the consul service
+// for 2 seconds and then asserts that the proxyid file gets written correctly.
+func TestRun_RetryServicePolling(t *testing.T) {
+	t.Parallel()
+	proxyFile := common.WriteTempFile(t, "")
+
+	// Start Consul server.
+	server, err := testutil.NewTestServerConfigT(t, nil)
+	defer server.Stop()
+	require.NoError(t, err)
+	server.WaitForLeader(t)
+	consulClient, err := api.NewClient(&api.Config{Address: server.HTTPAddr})
+	require.NoError(t, err)
+
+	ui := cli.NewMockUi()
+	cmd := Command{
+		UI:                                 ui,
+		consulClient:                       consulClient,
+		ProxyIDFile:                        proxyFile,
+		ServiceRegistrationPollingAttempts: 10,
+	}
+	// Start the command asynchronously, later registering the services.
+	exitChan := runCommandAsynchronously(&cmd, defaultTestFlags)
+	// Wait a moment.
+	time.Sleep(time.Second * 1)
+	// Register Consul services.
+	testConsulServices := []api.AgentServiceRegistration{consulCountingSvc, consulCountingSvcSidecar}
+	for _, svc := range testConsulServices {
+		require.NoError(t, consulClient.Agent().ServiceRegister(&svc))
+		time.Sleep(time.Second * 2)
+	}
+
+	// Assert that it exits cleanly or timeout.
+	select {
+	case exitCode := <-exitChan:
+		require.Equal(t, 0, exitCode)
+	case <-time.After(time.Second * 10):
+		// Fail if the stopCh was not caught.
+		require.Fail(t, "timeout waiting for command to exit")
```

Comment:
Yeah I'm with you on this one. Wondering how we can address it given the design of the code is such that it doesn't have signal handling because it isn't actually supposed to run in a go func like we're doing in this test?